src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -75,18 +76,60 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
       response.sendError(500, "An error occurred while fetching Youtube Posts");
       return;
     }
+
+    int currentPage = convertToInt(request.getParameter("currentPage"));
+    int postPerPage = convertToInt(request.getParameter("postPerPage"));
+
+    threadTitles.clear();
+    threadSentiments.clear();
+    threadLikes.clear();
+    threadUrls.clear();
     for (YoutubePost post : newPosts) {
       threadTitles.add(post.getTitle());
       threadSentiments.add(analyze.getSentimentScore(post.getContent()));
-      threadUpvotes.add(AnalyzedVideo.getRandomUpvote());
+      threadLikes.add(AnalyzedVideo.getRandomUpvote());
       threadUrls.add(post.getUrl());
+    }
 
-      threadInfoList.put(TITLE, threadTitles);
-      threadInfoList.put(SENTIMENT, threadSentiments);
-      threadInfoList.put(UPVOTES, threadUpvotes);
-      threadInfoList.put(URL, threadUrls);
+    int numOfPages = (int) Math.ceil(threadTitles.size() / postPerPage);
+    createCurrentPage(currentPage, postPerPage);
+    if (numOfPages == 0) {
+      numOfPages++;
     }
+    threadInfoList.put("numOfPages", numOfPages);
+    threadInfoList.put(TITLE, threadTitles);
+    threadInfoList.put(SENTIMENT, threadSentiments);
+    threadInfoList.put(LIKES, threadLikes);
+    threadInfoList.put(URL, threadUrls);
+
     response.setContentType("application/json;");
     response.getWriter().print(threadInfoList);
   }
+  private void createCurrentPage(int currentPage, int postPerPage) {
+    int start = (currentPage - 1) * postPerPage;
+    int end = currentPage * postPerPage;
+    if (threadTitles.size() < end) {
+      end = threadTitles.size();
+    }
+    threadTitles = threadTitles.subList(start, end);
+    threadSentiments = threadSentiments.subList(start, end);
+    threadLikes = threadLikes.subList(start, end);
+    threadUrls = threadUrls.subList(start, end);
+  }
+
+  private int convertToInt(String beingconverted) {
+    int convertee;
+    try {
+      convertee = Integer.parseInt(beingconverted);
+    } catch (NumberFormatException e) {
+      System.err.println("Could not convert to int: " + beingconverted);
+      return 1;
+    }
+
+    if (convertee < 1 || convertee > 10) {
```

Comment:
Does this condition still apply? If so, please throw an exception instead of returning 1 - currently, it's unclear to the caller if the function is returning a 1 because the input was valid or invalid.